TWTValidation/JSON Validator/Parser/TWTJSONSchemaParser.m

```diff
@@ -0,0 +1,652 @@
+    node.additionalPropertiesNode = [self parseAdditionalItemsOrPropertiesWithSchema:objectSchema forKey:kTWTJSONSchemaKeywordAdditionalProperties];
+    node.propertyDependencies = [self parseDependenciesWithSchema:objectSchema];
+
+    return node;
+}
+
+
+#pragma mark - Keyword and value parser methods
+
+- (TWTJSONSchemaASTNode *)parseCommonKeywordsFromSchema:(NSDictionary *)schema toNode:(TWTJSONSchemaASTNode *)node
+{
+    node.schemaTitle = [self parseStringFromSchema:schema forKey:kTWTJSONSchemaKeywordTitle];
+    node.schemaDescription = [self parseStringFromSchema:schema forKey:kTWTJSONSchemaKeywordDescription];
+    NSArray *validValues = [self parseArrayFromSchema:schema forKey:kTWTJSONSchemaKeywordEnum];
+    node.validValues = validValues ? [NSSet setWithArray:validValues] : nil;
+    node.andSchemas = [self parseArrayOfSchemasFromSchema:schema forKey:kTWTJSONSchemaKeywordAllOf];
+    node.orSchemas = [self parseArrayOfSchemasFromSchema:schema forKey:kTWTJSONSchemaKeywordAnyOf];
+    node.exactlyOneOfSchemas = [self parseArrayOfSchemasFromSchema:schema forKey:kTWTJSONSchemaKeywordOneOf];
+    node.notSchema = [self parseSchemaFromSchema:schema forKey:kTWTJSONSchemaKeywordNot];
+    node.definitions = [self parseDictionaryFromSchema:schema forKey:kTWTJSONSchemaKeywordDefinitions];
+
+    return node;
+}
+
+// Keyword: type
+// Example: { "type" : "boolean" }
+//          { "type" : [ "string", "object", "null" ] }
+- (NSSet *)parseTypeKeywordFromSchema:(NSDictionary *)schema
+{
+    id typeValue = schema[kTWTJSONSchemaKeywordType];
+
+    if (!typeValue) {
+        return nil;
+    }
+
+    // Relies on parseSchema: having filtered out schemas with type array, object, string, number, or integer. Thus, type is either:
+    // A) One of these strings: "boolean", "null", "any"
+    // B) An array of valid type strings
+    // C) Invalid wrt the JSON Schema definition of the type keyword
+
+    [self pushPathComponent:kTWTJSONSchemaKeywordType];
+
+    static NSSet *allowedTypes = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        allowedTypes = [NSSet setWithObjects:kTWTJSONSchemaTypeKeywordAny, kTWTJSONSchemaTypeKeywordBoolean, kTWTJSONSchemaTypeKeywordNull, nil];
+    });
+
+    if ([allowedTypes containsObject:typeValue]) {
+        [self popPathComponent];
+        return [NSSet setWithObject:typeValue];
+    }
+
+    [self checkObjectIsArrayWithAtLeastOneItem:typeValue allowNil:NO];
+
+    [typeValue enumerateObjectsUsingBlock:^(NSString *typeString, NSUInteger index, BOOL *stop) {
+        [self pushPathComponent:@(index)];
+        [self checkObject:typeString isOneOf:[self validJSONTypeKeywords]];
+        [self popPathComponent];
+    }];
+
+    [self popPathComponent];
+    return [NSSet setWithArray:typeValue];
+}
+
+
+// Keywords: exclusiveMinimum & Maximum, uniqueItems
+- (BOOL)parseBooleanValueFromSchema:(NSDictionary *)schema forKey:(NSString *)key valueIfNotPresent:(BOOL)defaultValue
```

Comment:
I think I prefer the key occurring before the schema in the method name. Something like, `parseBooleanForKey:schema:valueIfNotPresent:`. It reads a little more naturally, and I’m probably more interested in the key than the schema when I’m reading the code, right?